rfcs/20220201-quantization-scaling/README.md

```diff
@@ -0,0 +1,368 @@
+details.
+
+
+### 2.3. Scales for inputs in Concat
+
+The mechanism `attr:set_scales` will be introduced to Concat primitive. This
+mechanism is used in Binary primitive, so API is already there.
+
+The following limitations are in place for an initial implementation:
+ - no-runtime support;
+ - mask is 0.
+
+An example:
+
+```cpp
+    primitive_attr attr;
+    for (int n = 0; n < num_src; ++n)
+        attr.set_scales(DNNL_ARG_MULTIPLE_SRC + n, /* mask */ 0, src_scales[n]);
+```
+
+## 3. Unifying scaling in long-term
+
+The motivation behind unifying scaling:
+- Remove duplications;
+- Provide an aligned support across all use cases.
+
+There are a few issues with the current scaling support:
+ - Output scale, scales and binary_mul overlap each other;
+ - RNN quantization is specific to RNN, but can be replaced by other
+ scaling mechanisms:
+   ```
+   # Scaling semantics of set_scales and set_zero_points
+   x_f32 = scale * (x_i8 - zero_point)
+
+   # Scaling semantics of RNN quantization
+   x_i8 = scale * x_f32 + zero_point
+   x_f32 = 1 / scale * (x_i8 - zero_point)
+   ```
+   This will affect application side code as user should pass `1 / scale` to
+   oneDNN comparing to the current `scale`;
+ - Built-in parameter provides very limited flexibility (no mask, no runtime).
+
+### The changes:
+
+#### New quantization scheme
+
+```
+(dst - zp_dst) * scale_dst = post_ops(OP(...))
+```
+Notes:
+- scales/zero_points are applied to their respective argument
+  - To get dst output of post_ops() should be divided(!) on scale_dst
+- if a users needs rescaling/reshifting in the post chain, they can rely on binary post-op
+- Scales will be supported for all data types
+
+
+#### A unified scaling API using mechanism `attr::set_scales`
+
+- The encoding of `DNNL_ARG_SCALE` is changed to `512` to avoid overlapping with other arguments.
+- Mapping of existing scaling to `attr::set_scales`:
+  - `attr::output_scales`
+    ``` cpp
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_DST, mask, output_scales);
+    ```
+  - `attr::p_op::binary_mul`
+    ``` cpp
+     primitive_attr attr;
+     attr.set_scales(DNNL_ARG_DST, mask, output_scales);
+    ```
+  - `attr::rnn_data_qparams`
+    ``` cpp
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_SRC, mask, data_scales);
+    attr.set_scales(DNNL_ARG_DST, mask, data_scales);
+    ```
+  - `attr::rnn_weights_qparams`
+    ``` cpp
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_WEIGHTS, mask, weights_scales);
+    ```
+  - `attr::p_op::built-in parameter`
+    ``` cpp
+    post_ops p_ops;
+    p_ops.append_binary(algorithm::binary_add, summand_md);
+    p_ops.append_eltwise(algorithm::eltwise_relu, alpha, beta);
+
+    primitive_attr attr;
+    attr.set_post_ops(p_ops);
+    attr.set_scales(DNNL_ARG_ATTR_MULTIPLE_POST_OP(0) | DNNL_ARG_SRC_1,
+            /* mask */ 0, /* binary scale */ {DNNL_RUNTIME_F32_VAL});
+    attr.set_scales(DNNL_ARG_ATTR_MULTIPLE_POST_OP(1) | DNNL_ARG_DST,
```

Comment:
Do I understand correctly that `binary_mul` post_op after `eltwise_relu` actually provides the same math as `relu` dst scales?
In general different framework use different ways to express quantization information in their internal graphs. Some frameworks use separate opset for quantized operations which contains scales/shifts integrated in their semantics. In fact `attrs.set_scales` API perfectly fits such representation.
OpenVINO express quantization logic (quantize/dequantize/requantize) as separate "eltwise/binary" operations inside the graph. From that perspective it is more natural and convenient to pass scales/shits to oneDNN primitive along with other post_ops via unified logic without necessity to remap into special `set_scales` API.
The question: is there a guarantee that both ways to express quantization inside oneDNN will be supported with the similar performance by the backends?

Reply:
I see your point. I would say that expressing quantization via post operations will be more mature at the beginning, because this mechanism is already implemented using post_op injectors and we will continue using it. On the other hand we will add arg scales to standalone primitives and limited set of post operations (sum,binary_add) using optimized implementations (reusing existing post_op injectors for output scales would guarantee the same level of performance) and then build from there based on requests from users.